# ENGR-213 PSET 1
## IBM Financial Ratios (2022 vs 2023)

**Harsh Dadhich**

Looking at IBM's efficiency through two lenses:
- **Total Asset Turnover** — how well they use assets to generate revenue
- **Inventory Turnover** — how quickly they move through inventory

Goal: see if IBM got better or worse at these between 2022 and 2023.

## imports

In [21]:
import yfinance as yf
import pandas as pd

pd.set_option("display.float_format", lambda x: "%.2f" % x)

## fetching IBM data

In [22]:
# grab ibm ticker and pull financial statements
ibm = yf.Ticker("IBM")

balance_sheet = ibm.balance_sheet
income_statement = ibm.income_stmt

print("balance sheet:")
display(balance_sheet)

balance sheet:


,2024-12-31,2023-12-31,2022-12-31,2021-12-31
Treasury Shares Number,1352874243.00,1351897513.00,1351024943.00,1350509249.00
Ordinary Shares Number,926290070.00,915013646.00,906091977.00,898068599.00
Share Issued,2279164313.00,2266911159.00,2257116920.00,2248577848.00
Net Debt,41026000000.00,43479000000.00,43063000000.00,45053000000.00
Total Debt,58396000000.00,59935000000.00,54013000000.00,55139000000.00
...,...,...,...,...
Allowance For Doubtful Accounts Receivable,-114000000.00,-192000000.00,-233000000.00,-218000000.00
Gross Accounts Receivable,6918000000.00,7406000000.00,6774000000.00,6972000000.00
Cash Cash Equivalents And Short Term Investments,14591000000.00,13441000000.00,8738000000.00,7250000000.00
Other Short Term Investments,644000000.00,373000000.00,852000000.00,600000000.00


In [23]:
print("income statement:")
display(income_statement)

income statement:


,2024-12-31,2023-12-31,2022-12-31,2021-12-31
Tax Effect Of Unusual Items,-29610000.00,-67230000.00,29400000.00,3383915.65
Tax Rate For Calcs,0.21,0.14,0.21,0.03
Normalized EBITDA,12317000000.00,15191000000.00,7034000000.00,12277000000.00
Total Unusual Items,-141000000.00,-498000000.00,140000000.00,132000000.00
Total Unusual Items Excluding Goodwill,-141000000.00,-498000000.00,140000000.00,132000000.00
Net Income From Continuing Operation Net Minority Interest,6015000000.00,7514000000.00,1783000000.00,4712000000.00
Reconciled Depreciation,4667000000.00,4396000000.00,4802000000.00,6417000000.00
Reconciled Cost Of Revenue,23639000000.00,24159000000.00,27842000000.00,25865000000.00
EBITDA,12176000000.00,14693000000.00,7174000000.00,12409000000.00
EBIT,7509000000.00,10297000000.00,2372000000.00,5992000000.00


## extracting the numbers we need

from the balance sheet we need total assets and inventory (2021-2023 for avg calculations), and from income stmt we need revenue and COGS for 2022/2023.

In [24]:
# check what years we have available
print("years in balance sheet:")
for col in balance_sheet.columns:
    print(f"  {col.year}")

print("\nyears in income stmt:")
for col in income_statement.columns:
    print(f"  {col.year}")

years in balance sheet:
  2024
  2023
  2022
  2021

years in income stmt:
  2024
  2023
  2022
  2021


In [25]:
# helper to grab column by year (yfinance cols are datetime objs)
def get_col_by_year(df, year):
    for col in df.columns:
        if col.year == year:
            return col
    return None


# pull total assets 2021-2023
total_assets_2021 = balance_sheet.loc[
    "Total Assets", get_col_by_year(balance_sheet, 2021)
]
total_assets_2022 = balance_sheet.loc[
    "Total Assets", get_col_by_year(balance_sheet, 2022)
]
total_assets_2023 = balance_sheet.loc[
    "Total Assets", get_col_by_year(balance_sheet, 2023)
]

# pull inventory 2021-2023
inventory_2021 = balance_sheet.loc["Inventory", get_col_by_year(balance_sheet, 2021)]
inventory_2022 = balance_sheet.loc["Inventory", get_col_by_year(balance_sheet, 2022)]
inventory_2023 = balance_sheet.loc["Inventory", get_col_by_year(balance_sheet, 2023)]

# pull revenue 2022-2023
revenue_2022 = income_statement.loc[
    "Total Revenue", get_col_by_year(income_statement, 2022)
]
revenue_2023 = income_statement.loc[
    "Total Revenue", get_col_by_year(income_statement, 2023)
]

# pull cogs 2022-2023
cogs_2022 = income_statement.loc[
    "Cost Of Revenue", get_col_by_year(income_statement, 2022)
]
cogs_2023 = income_statement.loc[
    "Cost Of Revenue", get_col_by_year(income_statement, 2023)
]

# print extracted vals (in millions)
print("extracted data (millions USD):\n")
print(f"total assets:")
print(f"  2021: ${total_assets_2021/1e6:,.2f}M")
print(f"  2022: ${total_assets_2022/1e6:,.2f}M")
print(f"  2023: ${total_assets_2023/1e6:,.2f}M")

print(f"\ninventory:")
print(f"  2021: ${inventory_2021/1e6:,.2f}M")
print(f"  2022: ${inventory_2022/1e6:,.2f}M")
print(f"  2023: ${inventory_2023/1e6:,.2f}M")

print(f"\nrevenue:")
print(f"  2022: ${revenue_2022/1e6:,.2f}M")
print(f"  2023: ${revenue_2023/1e6:,.2f}M")

print(f"\ncogs:")
print(f"  2022: ${cogs_2022/1e6:,.2f}M")
print(f"  2023: ${cogs_2023/1e6:,.2f}M")

extracted data (millions USD):

total assets:
  2021: $132,001.00M
  2022: $127,243.00M
  2023: $135,241.00M

inventory:
  2021: $1,649.00M
  2022: $1,552.00M
  2023: $1,161.00M

revenue:
  2022: $60,530.00M
  2023: $61,860.00M

cogs:
  2022: $27,842.00M
  2023: $27,560.00M


## calculating the ratios

formulas:
- **total asset turnover** = revenue / avg total assets
- **inventory turnover** = cogs / avg inventory

In [26]:
# avg total assets (beginning + ending / 2)
avg_total_assets_2022 = (total_assets_2021 + total_assets_2022) / 2
avg_total_assets_2023 = (total_assets_2022 + total_assets_2023) / 2

# avg inventory
avg_inventory_2022 = (inventory_2021 + inventory_2022) / 2
avg_inventory_2023 = (inventory_2022 + inventory_2023) / 2

print("averages (millions USD):\n")
print(f"avg total assets:")
print(f"  2022: ${avg_total_assets_2022/1e6:,.2f}M")
print(f"  2023: ${avg_total_assets_2023/1e6:,.2f}M")

print(f"\navg inventory:")
print(f"  2022: ${avg_inventory_2022/1e6:,.2f}M")
print(f"  2023: ${avg_inventory_2023/1e6:,.2f}M")

averages (millions USD):

avg total assets:
  2022: $129,622.00M
  2023: $131,242.00M

avg inventory:
  2022: $1,600.50M
  2023: $1,356.50M


In [27]:
# total asset turnover = revenue / avg assets
total_asset_turnover_2022 = revenue_2022 / avg_total_assets_2022
total_asset_turnover_2023 = revenue_2023 / avg_total_assets_2023

# inventory turnover = cogs / avg inventory
inventory_turnover_2022 = cogs_2022 / avg_inventory_2022
inventory_turnover_2023 = cogs_2023 / avg_inventory_2023

print("results:\n")
print(f"total asset turnover:")
print(f"  2022: {total_asset_turnover_2022:.4f}")
print(f"  2023: {total_asset_turnover_2023:.4f}")

print(f"\ninventory turnover:")
print(f"  2022: {inventory_turnover_2022:.4f}")
print(f"  2023: {inventory_turnover_2023:.4f}")

results:

total asset turnover:
  2022: 0.4670
  2023: 0.4713

inventory turnover:
  2022: 17.3958
  2023: 20.3170


## comparing 2022 vs 2023

let's see how these changed year over year.

In [28]:
# yoy changes
tat_change = total_asset_turnover_2023 - total_asset_turnover_2022
tat_pct_change = (
    (total_asset_turnover_2023 - total_asset_turnover_2022) / total_asset_turnover_2022
) * 100

inv_change = inventory_turnover_2023 - inventory_turnover_2022
inv_pct_change = (
    (inventory_turnover_2023 - inventory_turnover_2022) / inventory_turnover_2022
) * 100

# summary table
comparison_data = {
    "Ratio": ["Total Asset Turnover", "Inventory Turnover"],
    "2022": [total_asset_turnover_2022, inventory_turnover_2022],
    "2023": [total_asset_turnover_2023, inventory_turnover_2023],
    "Change": [tat_change, inv_change],
    "% Change": [tat_pct_change, inv_pct_change],
    "Trend": [
        "Improved" if tat_change > 0 else "Worsened",
        "Improved" if inv_change > 0 else "Worsened",
    ],
}

comparison_df = pd.DataFrame(comparison_data).set_index("Ratio")
display(comparison_df)

,2022,2023,Change,% Change,Trend
Ratio,,,,,
Total Asset Turnover,0.47,0.47,0.00,0.94,Improved
Inventory Turnover,17.40,20.32,2.92,16.79,Improved


In [29]:
# breakdown
print("total asset turnover:")
print(f"  2022: {total_asset_turnover_2022:.4f}")
print(f"  2023: {total_asset_turnover_2023:.4f}")
print(f"  change: {tat_pct_change:+.2f}%")
print(f"  verdict: {'improved' if tat_change > 0 else 'worsened'}")

print(f"\ninventory turnover:")
print(f"  2022: {inventory_turnover_2022:.4f}")
print(f"  2023: {inventory_turnover_2023:.4f}")
print(f"  change: {inv_pct_change:+.2f}%")
print(f"  verdict: {'improved' if inv_change > 0 else 'worsened'}")

total asset turnover:
  2022: 0.4670
  2023: 0.4713
  change: +0.94%
  verdict: improved

inventory turnover:
  2022: 17.3958
  2023: 20.3170
  change: +16.79%
  verdict: improved


## Observations

**Total asset turnover** Went from 0.4670 to 0.4713 (+0.94%). Small uptick - IBM squeezed a bit more revenue out of each dollar in assets. not a dramatic shift, but moving in the right direction.

**Inventory turnover** jumped from 17.40 to 20.32 (+16.79%). This is the standout. IBM's moving through inventory ~17% faster than last year. they cut avg inventory from $1.6B to $1.36B while keeping cogs roughly flat - that's tighter ops.

**Bottom line:** Both metrics improved yoy. the inventory turnover gain is particularly strong and suggests IBM's gotten leaner on the supply chain side. asset turnover is relatively flat but stable - for a $135B asset base, that's not bad.

---

## Conclusion

| ratio | 2022 | 2023 | change |
|-------|------|------|--------|
| total asset turnover | 0.4670 | 0.4713 | +0.94% |
| inventory turnover | 17.40 | 20.32 | +16.79% |

Both ratios improved from 2022 to 2023, which is a good sign for IBM.

The **total asset turnover** barely moved (+0.94%), but it's still positive. IBM generates about $0.47 in revenue per dollar of assets, not amazing, but steady. for a company sitting on $135B in assets, holding the line is respectable.

The **inventory turnover** is where the story gets interesting. jumping from 17.4x to 20.3x means IBM is cycling through inventory about 3 more times per year. they trimmed avg inventory by ~$250M while keeping sales stable. that's leaner operations and better working capital mgmt.

**What does this mean for IBM?** They're getting more efficient, especially on the inventory side. the asset turnover staying flat isn't ideal, but the inventory gains suggest they're tightening up operations where they can. overall, these ratios paint a picture of a company that's improving its efficiency year over year - nothing explosive, but solid progress.

---
*Harsh Dadhich — engr-213 pset 1*